samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,747 @@
+	if (err) {
+		LOG_ERR("Cloud backend could not be initialized, error: %d", err);
+		return err;
+	}
+
+	/* Begin attempting to connect persistently. */
+	while (true) {
+		LOG_INF("Next connection retry in %d seconds",
+			CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS);
+
+		err = nrf_cloud_connect(NULL);
+		if (err) {
+			LOG_ERR("cloud_connect, error: %d", err);
+		}
+
+		/* Wait for cloud connection success. If succeessful, break out of the loop. */
+		if (await_cloud_connected(
+			K_SECONDS(CONFIG_CLOUD_CONNECTION_RETRY_TIMEOUT_SECONDS))) {
+			break;
+		}
+	}
+
+	/* Wait for cloud to become ready, resetting if we time out or are disconnected. */
+	if (!await_cloud_ready(K_SECONDS(CONFIG_CLOUD_READY_TIMEOUT_SECONDS), true)) {
+		LOG_INF("nRF Cloud failed to become ready. Resetting connection.");
+		reset_cloud();
+		return -ETIMEDOUT;
+	}
+
+	LOG_INF("Connected to nRF Cloud");
+	return 0;
+}
+
+
+/**
+ * @brief Set up modem and start trying to connect to LTE.
+ *
+ * @return int - 0 on success, otherwise a negative error code.
+ */
+static int setup_lte(void)
+{
+	int err;
+
+	LOG_INF("Setting up LTE");
+
+	/* Register to be notified when the Modem has figured out the current time. */
+	date_time_register_handler(date_time_evt_handler);
+
+	/* Perform Configuration */
+	if (IS_ENABLED(CONFIG_LTE_AUTO_INIT_AND_CONNECT)) {
+		/* Do nothing, modem is already configured and LTE connected. */
+		return 0;
+	}
+
+	if (IS_ENABLED(CONFIG_POWER_SAVING_MODE_ENABLE)) {
+		/* Requesting PSM before connecting allows the modem to inform
+		 * the network about our wish for certain PSM configuration
+		 * already in the connection procedure instead of in a separate
+		 * request after the connection is in place, which may be
+		 * rejected in some networks.
+		 */
+		err = lte_lc_psm_req(true);
+		if (err) {
+			LOG_ERR("Failed to set PSM parameters, error: %d", err);
+			return err;
+		}
+
+		LOG_INF("PSM mode requested");
+	}
+
+	/* Modem events must be enabled before we can receive them. */
+	err = lte_lc_modem_events_enable();
+	if (err) {
+		LOG_ERR("lte_lc_modem_events_enable failed, error: %d", err);
+		return err;
+	}
+
+	/* Init the modem, and start keeping an active connection.
+	 * Note that if connection is lost, the modem will automatically attempt to
+	 * re-establish it after this call.
+	 */
+	LOG_INF("Starting connection to LTE carrier...");
+	err = lte_lc_init_and_connect_async(lte_event_handler);
+	if (err) {
+		LOG_ERR("Modem could not be configured, error: %d", err);
+		return err;
+	}
+
+	return 0;
+}
+
+void manage_connection(void)
+{
+	/* Enable the modem and start trying to connect to LTE.
+	 * This is done once only, since the modem handles connection persistence then after.
+	 *
+	 * (Once we request connection, the modem will automatically try to reconnect whenever
+	 *  connection is lost).
+	 */
+
+	LOG_INF("Connecting to LTE Network. This may take several minutes...");
+	while (setup_lte()) {
```

Comment:
Ah! Is there a way to be certain about that? Are AT commands guaranteed only to fail for non-transient reasons?

Well, at any rate, I've done as you suggested and removed the while loop. Rather than reboot (and potentially reboot-loop), or halt, I have altered the sample to attempt to continue as though there had been no error.

Reply:
I checked what the code does and I couldn't really spot anything which could cause transient failures. If setup_lte() fails, most likely the app side can't communicate with the modem at all. And in that case retrying won't help.